### PR TITLE
Add experimental coref docs

### PR DESCRIPTION
**Pull request by polm**
## Description

This PR includes just the docs written in #7264, which now cover the components in https://github.com/explosion/spacy-experimental/pull/17. #7264 can be closed in favor of this PR. 

This should not be merged before https://github.com/explosion/spacy-experimental/pull/17.

One question is what version should be listed for the introduction of these components; currently it's 3.4.

(A few unrelated lines in the architecture docs were modified by prettier.)

### Types of change

docs update

## Checklist

- [x] I confirm that I have the right to submit this contribution under the project's MIT license.
- [ ] I ran the tests, and all new and existing tests passed.
- [x] My changes don't require a change to the documentation, or if they do, I've added all required information.

### CODE REVIEW

**Comment by svlandeg** [comment]
> One question is what version should be listed for the introduction of these components; currently it's 3.4.

It shouldn't list any spaCy version yet, until we actually port it over from experimental.

File: website/docs/api/coref.md [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+
+> #### Installation
+>
+> ```bash
+> $ pip install -U spacy-experimental
+> ```
+
+<Infobox title="Important note" variant="warning">
+
+This component not yet integrated into spaCy core, and is available via the
```
Comment (svlandeg) [inline website/docs/api/coref.md:20]:
```suggestion
This component is not yet integrated into spaCy core, and is available via the
```

File: website/docs/api/coref.md [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+extension package
+[`spacy-experimental`](https://github.com/explosion/spacy-transformers). It
```
Comment (svlandeg) [inline website/docs/api/coref.md:22]: We should say from which version onwards `coref` will be in `spacy-experimental`, so it's clear for users whether or not they'll need to upgrade.

File: website/docs/api/coref.md [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+exposes the component via entry points, so if you have the package installed,
```
Comment (svlandeg) [inline website/docs/api/coref.md:23]: Maybe we can link to the docs on "entry points"

**Comment by adrianeboyd** [comment]
Are we sure we want to put this in the main docs? There's no explanation of `spacy-experimental` or tags related to this?

File: website/docs/api/coref.md [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+| ------------------------------------------ | ------------------------------------------------------------------------------------------------------- |
+| `Doc.spans[prefix + "_" + cluster_number]` | One coreference cluster, represented as single-token spans. Cluster numbers start from 1. ~~SpanGroup~~ |
+
+## Config and implementation {#config}
+
+The default config is defined by the pipeline component factory and describes
+how the component should be configured. You can override its settings via the
+`config` argument on [`nlp.add_pipe`](/api/language#add_pipe) or in your
+[`config.cfg` for training](/usage/training#config). See the
+[model architectures](/api/architectures) documentation for details on the
```
Comment (svlandeg) [inline website/docs/api/coref.md:55]: Perhaps link directly to the correct subsection

File: website/docs/api/coref.md [hunk 5]
```diff
@@ -0,0 +1,355 @@
...
+> ```
+
+| Setting               | Description                                                                                                                              |
+| --------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- |
+| `model`               | The [`Model`](https://thinc.ai/docs/api-model) powering the pipeline component. Defaults to [Coref](/api/architectures#Coref). ~~Model~~ |
+| `span_cluster_prefix` | The prefix for the keys for clusters saved to `doc.spans`. Defaults to `coref_clusters`. ~~str~~                                         |
+
+```python
+%%GITHUB_SPACY/spacy/pipeline/coref.py
+```
```
Comment (svlandeg) [inline website/docs/api/coref.md:76]: These need to be removed/commented out for now

**Review by svlandeg — COMMENTED** [review]
Thanks, good idea to move this to a new PR for clarity :-)

Had a few more minor comments.

File: website/docs/api/coref.md [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+`nlp.add_pipe("coref")` will work out-of-the-box.
+
+</Infobox>
+
+A `CoreferenceResolver` component groups tokens into clusters that refer to the
+same thing. Clusters are represented as SpanGroups that start with a prefix
+(`coref_clusters` by default).
+
+A `CoreferenceResolver` component can be paired with a
+[`SpanPredictor`](/api/spanpredictor) to expand single tokens to spans.
```
[the remark this answers is not in the document]
Comment (polm) [reply website/docs/api/coref.md:34]: Yes, thanks for catching that!

**Comment by adrianeboyd** [comment]
I think it would be okay to redefine the `experimental` tag to mean `spacy-experimental`, but we should then keep it that way in the future rather than trying to use it for both uses.

File: website/docs/api/coref.md [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+network,
+[inferring missing shapes](https://thinc.ai/docs/usage-models#validation) and
+setting up the label scheme based on the data. This method is typically called
+by [`Language.initialize`](/api/language#initialize).
+
+> #### Example
+>
+> ```python
+> coref = nlp.add_pipe("coref")
+> coref.initialize(lambda: [], nlp=nlp)
```
Comment (adrianeboyd) [inline website/docs/api/coref.md:166]: Update this following #11221?

File: website/docs/api/span-resolver.md [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+network,
+[inferring missing shapes](https://thinc.ai/docs/usage-models#validation) and
+setting up the label scheme based on the data. This method is typically called
+by [`Language.initialize`](/api/language#initialize).
+
+> #### Example
+>
+> ```python
+> span_resolver = nlp.add_pipe("span_resolver")
+> span_resolver.initialize(lambda: [], nlp=nlp)
```
Comment (adrianeboyd) [inline website/docs/api/span-resolver.md:166]: Also following #11221.

**Comment by polm** [comment]
> I think it would be okay to redefine the experimental tag to mean spacy-experimental, but we should then keep it that way in the future rather than trying to use it for both uses.

In that case what should we do for features like rehearsal, that can't be split out into `spacy-experimental`?

**Comment by adrianeboyd** [comment]
That is a good point. I think having two separate `experimental` and `spacy-experimental` tags is going to be too confusing.

**Comment by polm** [comment]
How about we:

1. use `experimental` for both `spacy-experimental` and experimental features in core
2. make a point of using warning infoboxes or other techniques to signpost things in `spacy-experimental`?

I think the point of the `experimental` tag is to indicate that APIs may be unstable or it may have seen less real-world use, and that's true of features in core or `spacy-experimental`, so sharing the tag makes sense. 

While it is true that it can be confusing if someone misses the note that an experimental feature is in a separate package, I think that's the kind of temporary confusion that's easily fixed, rather than something more serious like breaking an existing application. If we're diligent about signposting things that require the extra library, then I think the small chance of confusion is a fair tradeoff for the benefit of having the docs all in one place with consistent formatting (and being searchable too).

**Comment by polm** [comment]
I'll put this in draft state until https://github.com/explosion/spacy-experimental/pull/17 has been merged.

**Comment by polm** [comment]
Test failure here is due to https://github.com/python/mypy/issues/13627. Since this PR contains no code, only docs, it should be safe to ignore.

File: website/docs/api/architectures.md [hunk 9]
```diff
@@ -920,5 +921,84 @@ A function that reads an existing `KnowledgeBase` from file.
...
+with only token-level clusters is acceptable.
+
+### spacy.Coref.v1 {#Coref tag="experimental"}
+
+> #### Example Config
+>
+> ```ini
+>
+> [model]
+> @architectures = "spacy.Coref.v1"
```
Comment (svlandeg) [inline website/docs/api/architectures.md:942]: This needs to be `spacy-experimental.Coref.v1`, right? Can you go through all docs to ensure it's consistent everywhere?

File: website/docs/api/architectures.md [hunk 10]
```diff
@@ -920,5 +921,84 @@ A function that reads an existing `KnowledgeBase` from file.
...
+> pooling = {"@layers":"reduce_mean.v1"}
+> ```
+
+The `SpanResolver` model architecture is a Thinc `Model`.
+
+| Name                      | Description                                                                                                                   |
+| ------------------------- | ----------------------------------------------------------------------------------------------------------------------------- |
+| `tok2vec`                 | The [`tok2vec`](#tok2vec) layer of the model. ~~Model~~                                                                       |
+| `distance_embedding_size` | A representation of the distance between two candidates. ~~int~~                                                              |
+| `dropout`                 | The dropout to use internally. Unlike some Thinc models, this has separate dropout for the internal PyTorch layers. ~~float~~ |
```
Comment (svlandeg) [inline website/docs/api/architectures.md:998]: `dropout` isn't a parameter of `SpanResolver`, or is it?

File: website/docs/api/coref.md [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+how the component should be configured. You can override its settings via the
+`config` argument on [`nlp.add_pipe`](/api/language#add_pipe) or in your
+[`config.cfg` for training](/usage/training#config). See the
+[model architectures](/api/architectures#coref-architectures) documentation for
+details on the architectures and their arguments and hyperparameters.
+
+> #### Example
+>
+> ```python
+> from spacy.pipeline.coref import DEFAULT_COREF_MODEL
```
Comment (svlandeg) [inline website/docs/api/coref.md:62]: I assume this code won't work anymore?

File: website/docs/api/coref.md [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+> ```python
+> # Construction via add_pipe with default model
+> coref = nlp.add_pipe("coref")
+>
+> # Construction via add_pipe with custom model
+> config = {"model": {"@architectures": "my_coref.v1"}}
+> coref = nlp.add_pipe("coref", config=config)
+>
+> # Construction from class
+> from spacy.pipeline import CoreferenceResolver
```
Comment (svlandeg) [inline website/docs/api/coref.md:88]: Code won't work

File: website/docs/api/span-resolver.md [hunk 13]
```diff
@@ -0,0 +1,353 @@
...
+installed, using `factory = "span_resolver"` in your
+[training config](/usage/training#config) or `nlp.add_pipe("span_resolver")`
+will work out-of-the-box.
+
+</Infobox>
+
+A `SpanResolver` component takes in tokens (represented as `Span`s of length 1)
+and resolves them into `Span`s of arbitrary length. The initial use case is as a
+post-processing step on word-level [coreference resolution](/api/coref). The
+input and output keys used to store `Span`s are configurable.
```
Comment (svlandeg) [inline website/docs/api/span-resolver.md:33]: To avoid the slightly awkward "`Span`'s" I'd prefer something like "`Span` objects"

File: website/docs/api/span-resolver.md [hunk 14]
```diff
@@ -0,0 +1,353 @@
...
+how the component should be configured. You can override its settings via the
+`config` argument on [`nlp.add_pipe`](/api/language#add_pipe) or in your
+[`config.cfg` for training](/usage/training#config). See the
+[model architectures](/api/architectures#coref-architectures) documentation for
+details on the architectures and their arguments and hyperparameters.
+
+> #### Example
+>
+> ```python
+> from spacy.pipeline.span_resolver import DEFAULT_span_resolver_MODEL
```
Comment (svlandeg) [inline website/docs/api/span-resolver.md:60]: TODO: update

File: website/docs/api/span-resolver.md [hunk 15]
```diff
@@ -0,0 +1,353 @@
...
+> ```python
+> # Construction via add_pipe with default model
+> span_resolver = nlp.add_pipe("span_resolver")
+>
+> # Construction via add_pipe with custom model
+> config = {"model": {"@architectures": "my_span_resolver.v1"}}
+> span_resolver = nlp.add_pipe("span_resolver", config=config)
+>
+> # Construction from class
+> from spacy.pipeline import SpanResolver
```
Comment (svlandeg) [inline website/docs/api/span-resolver.md:87]: TODO: update

File: website/docs/api/architectures.md (hunk 10 above)
Comment (polm) [reply website/docs/api/architectures.md]: It's not, this must have been copied from the other component. Removed.

File: website/docs/api/span-resolver.md (hunk 13 above)
Comment (polm) [reply website/docs/api/span-resolver.md]: Sounds good, changed.

**Comment by polm** [comment]
It looks like this hadn't actually been updated to make sure the sample code worked since the implementation moved to `spacy-experimental`, but that should be taken care of now.

File: website/docs/api/architectures.md [hunk 16]
```diff
@@ -920,5 +921,84 @@ A function that reads an existing `KnowledgeBase` from file.
...
+
+## Coreference {#coref-architectures tag="experimental"}
+
+A [`CoreferenceResolver`](/api/coref) component identifies tokens that refer to
+the same entity. A [`SpanResolver`](/api/span-resolver) component infers spans
+from single tokens. Together these components can be used to reproduce
+traditional coreference models. You can also omit the `SpanResolver` if working
+with only token-level clusters is acceptable.
+
+### spacy.Coref.v1 {#Coref tag="experimental"}
```
Comment (svlandeg) [inline website/docs/api/architectures.md:935]: TODO update

File: website/docs/api/architectures.md [hunk 17]
```diff
@@ -920,5 +921,84 @@ A function that reads an existing `KnowledgeBase` from file.
...
+| `hidden_size`             | Size of the main internal layers. ~~int~~                                                                                                                                                |
+| `depth`                   | Depth of the internal network. ~~int~~                                                                                                                                                   |
+| `antecedent_limit`        | How many candidate antecedents to keep after rough scoring. This has a significant effect on memory usage. Typical values would be 50 to 200, or higher for very long documents. ~~int~~ |
+| `antecedent_batch_size`   | Internal batch size. ~~int~~                                                                                                                                                             |
+| **CREATES**               | The model using the architecture. ~~Model[List[Doc], Floats2d]~~                                                                                                                         |
+
+### spacy.SpanResolver.v1 {#SpanResolver tag="experimental"}
```
Comment (svlandeg) [inline website/docs/api/architectures.md:970]: TODO update

**Review by svlandeg — COMMENTED** [review]
Found two more outdated spots:

File: website/docs/api/coref.md [hunk 18]
```diff
@@ -0,0 +1,353 @@
...
+>
+> ```bash
+> $ pip install -U spacy-experimental
+> ```
+
+<Infobox title="Important note" variant="warning">
+
+This component is not yet integrated into spaCy core, and is available via the
+extension package
+[`spacy-experimental`](https://github.com/explosion/spacy-transformers) starting
```
Comment (radandreicristian) [inline website/docs/api/coref.md:21]: I think the link here is wrong.

Should it be https://github.com/explosion/spacy-experimental instead?

Comment (polm) [reply website/docs/api/coref.md]: You're right, thanks for catching that!

File: website/docs/api/pipeline-functions.md [hunk 19]
```diff
@@ -153,3 +153,36 @@ whole pipeline has run.
...
+
+Remove `SpanGroup`s from `doc.spans` based on a key prefix. This is used to
+clean up after the [`CoreferenceResolver`](/api/coref) when it's paired with a
+[`SpanResolver`](/api/span-resolver).
+
+<Infobox title="Important note" variant="warning">
+
+This pipeline function is not yet integrated into spaCy core, and is available
+via the extension package
+[`spacy-experimental`](https://github.com/explosion/spacy-transformers) starting
```
Comment (svlandeg) [inline website/docs/api/pipeline-functions.md:167]: To fix

Comment (polm) [reply website/docs/api/pipeline-functions.md:167]: Gah! I found one other and thought I'd gotten them all. Thanks.
